vmm/src/lib.rs

```diff
@@ -491,17 +497,28 @@ impl EpollContext {
         (dispatch_base, sender)
     }
 
+    // See the below comment for `allocate_virtio_net_tokens`, for an explanation on the returned
+    // values.
     fn allocate_virtio_block_tokens(&mut self) -> (virtio::block::EpollConfig, usize) {
         let (dispatch_base, sender) = self.allocate_tokens(virtio::block::BLOCK_EVENTS_COUNT);
         (
             virtio::block::EpollConfig::new(dispatch_base, self.epoll_raw_fd, sender),
-            self.device_handlers.len(),
+            self.device_handlers.len() - 1,
```

Comment:
Why `-1`?

Reply:
The `allocate...` function returns a tuple that contains the newly allocated handler's ID. That's the last element of `self.device_handlers`, so `self.device_handlers.len() - 1`. If it were to return the length (sans the `-1`), as it did before, the caller would have to be privy to the functional logic of `EpollContext`, in order for it to deduce the handler ID. I.e. the caller would have to know that `EpollContext` always places new handlers at the end of its list. It make sense to encapsulate logic, to get weaker dependencies between objects, and easier future changes.